protocol/daemons/liquidation/client/sub_task_runner.go

```diff
@@ -0,0 +1,291 @@
+package client
+
+import (
+	"context"
+	gometrics "github.com/armon/go-metrics"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"github.com/dydxprotocol/v4-chain/protocol/daemons/flags"
+	"github.com/dydxprotocol/v4-chain/protocol/daemons/liquidation/api"
+	"github.com/dydxprotocol/v4-chain/protocol/lib"
+	"github.com/dydxprotocol/v4-chain/protocol/lib/metrics"
+	clobtypes "github.com/dydxprotocol/v4-chain/protocol/x/clob/types"
+	satypes "github.com/dydxprotocol/v4-chain/protocol/x/subaccounts/types"
+	"time"
+)
+
+// SubTaskRunner provides an interface that encapsulates the liquidations daemon logic to gather and report
+// potentially liquidatable subaccount ids. This interface is used to mock the daemon logic in tests.
+type SubTaskRunner interface {
+	RunLiquidationDaemonTaskLoop(
+		client *Client,
+		ctx context.Context,
+		liqFlags flags.LiquidationFlags,
+		subaccountQueryClient satypes.QueryClient,
+		clobQueryClient clobtypes.QueryClient,
+		liquidationServiceClient api.LiquidationServiceClient,
+	) error
+}
+
+type SubTaskRunnerImpl struct{}
+
+// Ensure SubTaskRunnerImpl implements the SubTaskRunner interface.
+var _ SubTaskRunner = (*SubTaskRunnerImpl)(nil)
+
+// RunLiquidationDaemonTaskLoop contains the logic to communicate with various gRPC services
+// to find the liquidatable subaccount ids.
+func (s *SubTaskRunnerImpl) RunLiquidationDaemonTaskLoop(
+	client *Client,
+	ctx context.Context,
+	liqFlags flags.LiquidationFlags,
+	subaccountQueryClient satypes.QueryClient,
+	clobQueryClient clobtypes.QueryClient,
+	liquidationServiceClient api.LiquidationServiceClient,
+) error {
+	defer telemetry.ModuleMeasureSince(
+		metrics.LiquidationDaemon,
+		time.Now(),
+		metrics.MainTaskLoop,
+		metrics.Latency,
+	)
+
+	// 1. Fetch all subaccounts from query service.
+	subaccounts, err := GetAllSubaccounts(
+		client,
+		ctx,
+		subaccountQueryClient,
+		liqFlags.SubaccountPageLimit,
+	)
+	if err != nil {
+		return err
+	}
+
+	// 2. Check collateralization statuses of subaccounts with at least one open position.
+	liquidatableSubaccountIds, err := GetLiquidatableSubaccountIds(
+		client,
+		ctx,
+		clobQueryClient,
+		liqFlags,
+		subaccounts,
+	)
+	if err != nil {
+		return err
+	}
+
+	// 3. Send the list of liquidatable subaccount ids to the daemon server.
+	err = SendLiquidatableSubaccountIds(
+		ctx,
+		liquidationServiceClient,
+		liquidatableSubaccountIds,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CheckCollateralizationForSubaccounts queries a gRPC server using `AreSubaccountsLiquidatable`
+// and returns a list of collateralization statuses for the given list of subaccount ids.
+func CheckCollateralizationForSubaccounts(
```

Comment:
why not make this a method with a receiver type?

```suggestion
func (c *Client) CheckCollateralizationForSubaccounts(
```

Reply:
My thought was, I don't really want to expose these methods to the app, which stores a reference to the client, since it's an implementation detail. But I'm open to it if you prefer it and think it's better.